### ISSUE TITLE
"Raw" byte slice command

### ISSUE THREAD

**Issue by wyndhblb** [redis/go-redis#472]
Under large load (many thousands of commands/second), the command -> string interface of the commands incurs huge GC issues (lots and lots of string copies that need collecting).   For 95% of uses this is ok, but that this load, a better solution is to keep things as byte slices and allow for those slices to get directly piped into either the pipeline buffer or sent to the pool connection directly.  As a start one can simply introduce a "raw" function that simply copies a byte slice to the buffer rather then creating the command from arguments. 

For example rather then 

conn.ZAdd(key, redis.Z{Score: Number), Member: MyKey})

allow for 

conn.Raw(mySlice) 

where the mySlice is composed before hand (usually using sync.Pool byte buffers and custom custom optimized encoders).  This will be particularly good for pipelines

**Comment by vmihailenco** [comment]
I am not sure I fully understand what you are proposing. `conn.ZAdd(key, redis.Z{Score: Number, Member: MyKey})` allocates like 100 bytes and that does not involve copying strings. I don't think that `conn.Raw` will make huge difference.

https://github.com/go-redis/redis/pull/475 should reduce number of allocations in some cases, but it will not make a huge difference too.

>the command -> string interface of the commands incurs huge GC issues (lots and lots of string copies that need collecting)

Can you please point me to these allocations?

**Comment by wyndhblb** [comment]
Hi .. when doing a 10k/s command into redis, the GC went insane (basically 50% of all CPU usage was GC mark/collecting).  I pprof traced the issue to this module where it was taking ~40% of all allocated memory. Basically the serialization of the redis CMDs objects down to "strings" then down  to bytes.  In normal use cases, this is probably not a issue.   

ZAdd is just what i was doing, (but i did a few other commands w/ the same result Sadd, Set, etc) imagine any commands suffer from the same issue at that kind of volume.  So i was just proposing that there is a mechanism to bypass the cmd serialization process and insert basically a []byte  into a [][]byte command list.  Or direct to the base io.Writer of the socket buffer itself.

Basically i had to revert to raw TCP socket io.Writer in order to not have this issue (but in clustering mode, or ring mode, things need more fancy logic.  

Another suggestion is to get that io.Writer from the module connection itself so the pooling and cluster/ring/sentinel goodies are still in play.

**Comment by vmihailenco** [comment]
I see 3 types of allocations there:
- slice of interfaces for `_args []interface{}`
- interface{} itself via runtime.convT2E
- memory for `Cmd` struct

All those allocations are very small `<= 64bytes` and I don't see how it is possible to avoid them. It is also hard to imagine the type of program that will suffer from such allocations, but obviously you have just such a program...

I will try to add `RawCmd`, but it looks like a micro-optimization to me. You probably just need a bigger / more servers to run your app...

**Comment by dim** [comment]
We have a similar application while Go's GC pauses are improving from release to release, they also out

**Comment by dim** [comment]
Sorry, writing on my phone, let's try this again :)

We have a similar application while Go's GC pauses are improving from release to release, they also put additional pressure on CPU resources. I too toying with a RawCmd idea, but never had the time to look into it properly.

**Comment by vmihailenco** [comment]
`RawCmd` will require exporting some private API to get cmd name and first key:

```
func NewRawCmd(name, firstKey string, payload []byte) {}
```

That will avoid some overhead at the cost that user has to write all the serialization code himself, which sounds almost like requiring user to use assembler...

Another idea would be to move `[]interface{}` allocation into command struct like this:

```
type baseCmd struct {
    bootstrap [10]interface{}
    _args []interface{}
}
```

and allocate most commands from `bootstrap`:

```
cmd._args = cmd.bootstrap[:4]
```

That will slightly increase memory usage, but decrease number of allocations. And hopefully help GC.

**Comment by vmihailenco** [comment]
>That will slightly increase memory usage, but decrease number of allocations. And hopefully help GC.

I just tried and it removes exactly 1 allocation by increasing memory usage by almost 90 bytes. Does not look too impressing...

**Comment by vmihailenco** [comment]
@wyndhblb I wonder what is the performance difference between go-redis & raw TCP socket in your case?

The only idea that seems worth a try is to add `WriteBuffer` to `Cmd`:

```
cmd := redis.NewStringCmd()
wb := cmd.WriteBuffer() // uses sync.Pool to allocate buffer
wb.AppendArgsNum(3)
wb.AppendString("set")
wb.AppendString("foo")
wb.AppendBytes([]byte("bar"))
client.Process(cmd) // returns buffer to sync.Pool
```

The only dubious part is using sync.Pool to allocate `WriteBuffer` which looks like a manual memory management...

**Comment by vmihailenco** [comment]
Closing since there is no much interest in this.

**Comment by ileyd** [comment]
This may also be used with redis modules that have no client support, such as rejson.

**Comment by vmihailenco** [comment]
I don't see how this can be used with rejson, but I don't have a plan for rejson so maybe...

**Comment by oudwud** [comment]
I also really want to get `[]byte` from `Get()`. In many cases, we store byte arrays instead of string. For example, I use Redis as a cache for the gRPC client. It means I use the unmarshalled protobuf message as Redis key/value.
If I call `[]byte(stringFromGet)` whenever I call `Get()`, it will definitely cause lots of GCs.
Also, it would be so useful to use `[]byte` as a key for `Get()` and `Set()`. 
Please consider this new feature. It will improve performance a lot.

**Comment by vmihailenco** [comment]
There is already `rdb.Get("foo").Bytes()` which I've just made faster in https://github.com/go-redis/redis/pull/1106. We can also add `SetBytes` to save 1 `interface{}` allocation but I wonder if that is measurable...

**Comment by oudwud** [comment]
> There is already `rdb.Get("foo").Bytes()` which I've just made faster in #1106. We can also add `SetBytes` to save 1 `interface{}` allocation but I wonder if that is measurable...

Thank you for your reply. I didn’t know you’ve used `unsafe` for performance. I think it is enough for now. Thank you.
